[FIX] rating: allow portal user to post rating

seb-odoo opened the pull request:
Follow up of https://github.com/odoo/odoo/pull/176203

Issue: any posting of rating as portal user on a thread with read access for posting. The simplest use case is posting a review on a product (website_sale) after the feature is enabled.

When writing `rating_ids` on message, it actually writes `message_id` on rating, however the portal user doesn't have access to this model.

seb-odoo commented on addons/rating/models/mail_thread.py:
```diff
@@ -23,7 +23,45 @@ def unlink(self):
 
     def _get_message_create_valid_field_names(self):
-        return super()._get_message_create_valid_field_names() | {'rating_ids'}
+        return super()._get_message_create_valid_field_names() | {"rating_id", "rating_value"}
```
Note: there is a lot of code for `rating_id` while it seems almost never used.
Main call is livechat `_post_feedback_message` which could be refactored to pass `rating_value` most of the time.
And changing rating from message doesn't seem very useful. It should probably be prevented, but each model kind of does its own:
- there is a check on livechat route, but it is not enforced in models
- there is a different check for website slides
- I've not checked others

Overall rating should be fully cleaned, just doing the minimal here to keep features intact.

seb-odoo commented on addons/rating/models/mail_thread.py:
```diff
@@ -23,7 +23,45 @@ def unlink(self):
+                    "res_id": self.id,
+                    "res_model_id": self.env["ir.model"]._get_id(self._name),
+                }
+                # sudo: rating.rating - current user can create a rating when posting a message
+                self.env["rating.rating"].sudo().create(rating_vals)
```
Note: this effectively means (even before these PR) that any new message can be created together with a rating. It seems like we would like more control over that, probably at the model level.

The options I see should be: free create (rare? or non-existent?), one per user per record (most of the time: product, slides, etc), only at the end of flows (livechat, task,etc).

tsm-odoo commented on addons/rating/models/mail_thread.py:
```diff
@@ -23,7 +23,45 @@ def unlink(self):
+    def _message_create(self, values_list):
+        """Override of _message_create rather than message_post to be just after create
+        but before any post-processing."""
+        rating_id_by_message_index = (values.pop("rating_id", None) for values in values_list)
+        rating_value_by_message_index = (values.pop("rating_value", None) for values in values_list)
```
Couldn't those two generators+the use of zip just be reduced to `enumerate` and direct access to value list? It looks overcomplicated?

```py
for i, message in enumerate(messages):
    rating = values_list[i].pop("rating_id", None)
    rating_value = values_list[i].pop("rating_value", None)
```

seb-odoo replied:
They need to be popped before calling super and are used afterwards.

tsm-odoo reviewed (commented):
What flow is it fixing? Rating from live chat seems to work fine. Could you add steps to reproduce this issue on your commit message?

seb-odoo commented:
Any posting of rating as portal user on a thread with read access for posting. The simplest use case is posting a review on a product (website_sale) after the feature is enabled.

tsm-odoo reviewed (approved):
@robodoo r+

tde-banana-odoo commented:
@robodoo r-

tde-banana-odoo commented:
Strongly disagree with this PR. message_post is the official API managing stuff like converting ratings, ensuring access rights, ... ``_message_create`` is a help that should never contain magic.

Don't have time to dive into it but I bet there are cleaner fixes.

seb-odoo commented:
I'll sudo message post from controller then, I wanted to avoid it conceptually, but I think we're fine doing it all the time, as it was already done most of the time and the security checks are already in place.

Still need to add a check that given `rating_id` is matching current user/thread though, as that was completely missing since like forever. But this can be done inside current code in message post.

tde-banana-odoo commented:
If you want to cleanup rating code, feel free to propose PR targetting only that code so that we can discuss calmly and write clean tests. Otherwise you just end up merging a bunch of code that makes no sense as you mix several completely unrelated features.

I am open to cleanup rating (it is planned and we have to do it) but that should be done completely independently from any discuss related cleaning.

Cheers !

seb-odoo commented on addons/rating/models/mail_thread.py:
```diff
@@ -166,8 +166,6 @@ def rating_apply(self, rate, token=None, rating=None, feedback=None,
         rating_id = kwargs.pop('rating_id', False)
         rating_value = kwargs.pop('rating_value', False)
-        # create rating.rating record linked to given rating_value. Using sudo as portal users may have
-        # rights to create messages and therefore ratings (security should be checked beforehand)
```
Note: this sudo/comment made no sense, as `message_post` is a public method there is no such things as "checking beforehand".

tde-banana-odoo replied:
When creating the message, it checks access on related record. So the wording is a bit misleading but yes checks are done.

seb-odoo commented on addons/rating/models/mail_thread.py:
```diff
@@ -177,7 +175,7 @@ def message_post(self, **kwargs):
                 'consumed': True,
                 'partner_id': self.env.user.partner_id.id,
             }
-            rating_id = self.env["rating.rating"].sudo().create(rating_vals).id
```
`message_post` itself now sudo from controller.

seb-odoo reviewed (commented):
Updated without using `_message_create`. Let me know if that is better for you.

tde-banana-odoo commented:
I don't understand this fix. Why removing a sudo at low level to add one top level ?

seb-odoo commented:
The one at top level is necessary to link message to rating when the message is created (the crashing line is no longer in diff, `kwargs["rating_ids"] = [Command.set(rating_id)]`).

Without the top-level sudo, calling super with this param set will raise with portal user on a thread with mail_post set as read: just run the test without the sudo to see the issue, or try to post a review on a product page as portal user.

Note:
- Internal users have the rights on rating so it doesn't raise for them before the fix.
- And portal user without natural access to thread (when they are posting using access token) don't have the issue either, because `thread` is already sudo for them in that case (all the security is already in place).

The fix is to always sudo in controller when access is allowed, as opposed to only doing it when access is allowed through token. This is more consistent and fixes the issue.

--- 

Then because there is always a top level sudo, the other lower-level sudo can be removed, unless we want to keep supporting portal user creating rating using message_post rpc directly in all circumstances? That doesn't seem useful.

tde-banana-odoo commented:
But it was intended to prevent from doing that. You removed a check about "rating_ids" that was done on purpose. Moreover we are going to remove all ACLs on rating model, so it makes no sense to keep the "rating_ids" support you introduced, we will have to remove it again.

seb-odoo commented on addons/im_livechat/controllers/main.py:
```diff
@@ -195,14 +195,14 @@ def get_session(self, channel_id, anonymous_name, previous_operator_id=None, cha
-            'rating_url': rating.rating_image_url,
-            'reason': reason,
+        body = Markup(
+            """<div class="o_mail_notification o_hide_author">"""
```
Note: only "cosmetic" change, since we now assert this value in test, it was obvious how ugly it was (many new lines and a lot of unnecessary space in body), so the opportunity is taken to simplify it.

seb-odoo commented on addons/im_livechat/models/discuss_channel_member.py:
```diff
@@ -26,7 +26,7 @@ def _gc_unpin_livechat_sessions(self):
         super()._to_store(store, **kwargs)
         for member in self.filtered(lambda m: m.channel_id.channel_type == "livechat"):
-            # sudo: mail.channel - reading livechat channel to check whether current member is a bot is allowed
+            # sudo: discuss.channel - reading livechat channel to check whether current member is a bot is allowed
```
Renamed since a long time ago, still sometimes doing typos by habits.

seb-odoo commented:
@tde-banana-odoo updated PR to take into account all discussions and relevant parts from your PR.

- no override of message create
- rating_ids not allowed
- rating_id not allowed directly, ignored and processed in message post after hook (it means it's not supported from message notify and log batch)
- create a new specific test model
- ensure it works both from controller or from direct call
- added basic security check when linking rating id

I added tests to cover both issues.

Let me know if we can move forward with this fix.

seb-odoo commented on addons/mail/models/mail_thread.py:
```diff
@@ -2873,6 +2873,14 @@ def _message_compute_subject(self):
     def _message_create(self, values_list):
         """ Low-level helper to create mail.message records. It is mainly used
         to hide the cleanup of given values, for mail gateway or helpers."""
+        values_list = [
```
Is it even possible to have `from`, `to`, `cc`?
It seems like they should raise in `_raise_for_invalid_parameters` as they are not in `_get_message_create_valid_field_names`?
If we need to support them somehow, then I could simply move them to the ignore list.

That's also why I cleaned up the ignore list at the top, it was to be before the raise for invalid.
Otherwise `ValueError: Those values are not supported when posting or notifying: rating_id`.

seb-odoo replied:
All these keys are failing too:
`ValueError: Those values are not supported when posting or notifying: to`

Is it dead code that I can remove, or I should add them to ignore list?

tde-banana-odoo replied:
It was related to mailgateway code (sending a bit too much values when injecting emails for message_post), not sure if still relevant or not, don't have much time to check that currently to be honest ^^ .

tde-banana-odoo replied:
Ah the pop is done after the check ? Ah indeed then it is dead code :D

seb-odoo commented on addons/im_livechat/tests/test_message.py:
```diff
@@ -144,3 +150,124 @@ def test_message_to_store(self):
+            )["discuss.channel"][0]["id"]
+        )
+
+        def _get_feedback_bus():
```
This test is not specifically about rating now, it's generally speaking having a coverage of the (full) feedback message as we had no coverage of it before.

It's nice to see it's correctly posted as a comment and not a note, that it doesn't have attachments, etc. It's not specifically testing the formatter, more the actual values in this case.

Also technically we're lacking tools to make partial diff on `assertBus`, not that they are especially hard to implement, just out of scope and low value to have.
